### PR TITLE
Create Generic DamageOnInteract/Attacked Comps/Systems

### PR DESCRIPTION
**Pull request by Plykiya**
## About the PR

Lightbulb and anomaly contact damage is pretty hardcoded. Changed to use generic systems.
Only like six gloves actually had light bulb protection, changed it so that all gloves get protection as well as hardsuits/EVA suits. Fingerless gloves do not get protection.

Borgs don't get burned when touching the lights anymore. (They shouldn't even have free hands to begin with, but that's a different issue.)

## Why / Balance

Fixes https://github.com/space-wizards/space-station-14/issues/29502

GloveHeatResistanceComponent needed to be removed, it was only used for checking if light bulbs hurt you. AnomalyContact damage was just something that came to mind that probably also didn't need to exist.

With the two new generic systems I can add in things like touching the hot grill with your bare hands hurting you without feeling gross about hardcoding it.

## Technical details

- Removed `GloveHeatResistanceComponent.cs`, it was only responsible for determining if light bulbs hurt you.
- Removed damage from `PowerLightComponent.cs` now that damage is declared in its `DamageOnInteract` component.
- Removed `PowerLightSystem.cs` logic relating to hot bulbs cancelling interactions, it's handled in `DamageOnInteract` system now.
- Removed `ContactDamage` and `ContactDamage` related interactions from Anomalies, they are handled by `DamageOnInteract` and `DamageOnAttacked` system now.
- Created the `DamageOnAttacked` and `DamageOnInteract` component and system. Interactions can be cancelled if the user does not have sufficient protection, attacks don't need to be canceled.
- Both systems support localized popup messages and changing the sound played when the user takes damage. Damage is determined by using DamageSpecifiers. Protection uses DamageModifierSet values. Damage - Protection = Resulting Damage.
- `TryGetInventoryEntity<T>` in `InventorySystem.Slots.cs` was updated to provide the `Entity<T>` rather than just the EntityUid. This isn't a breaking change, but it does conveniently give me a way to access the DamageProtection components that I'm checking for.

## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

https://github.com/user-attachments/assets/8d7ec147-90ba-4497-b5d3-7b06d02a778b

https://github.com/user-attachments/assets/c224fd20-e47f-4872-b309-80663901904a

## Breaking changes

- Removed `GloveHeatResistanceComponent.cs`, please use the `DamageOnInteract` and `DamageOnInteractProtection` components for determining damage.
- Removed damage from `PowerLightComponent.cs`, please use the `DamageOnInteract` component.
- Removed `ContactDamage` and `ContactDamage` related interactions from Anomalies, please use the `DamageOnInteract` and `DamageOnAttacked` components for determining damage.

**Changelog**

:cl:
- add: Hardsuits and EVA suits now count as protection for unscrewing lightbulbs.
- add: More gloves were given the ability to unscrew light bulbs.
- remove: Behonkers no longer hurt you when melee attacking them or interacting with them.

### CODE REVIEW

**Comment by Plykiya** [comment]
ah, right, forgot
it technically will deal damage now without power, I'll uh, deal with that

**Comment by Plykiya** [comment]
also forgot to give protection to entities like borgs which was the whole reason I made this PR, zzz I'm too tired and hungry to remember all the details right

File: Content.Shared/Damage/Components/DamageOnInteractComponent.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    /// and the <see cref="DamageOnInteractProtectionComponent"/>
+    /// </summary>
+    [DataField]
+    public bool IgnoreResistances = false;
+
+    /// <summary>
+    /// What kind of localized text should pop up when they interact with the entity
+    /// </summary>
+    [DataField]
+    public string? PopupText;
```
Comment (Tayrtahn) [inline Content.Shared/Damage/Components/DamageOnInteractComponent.cs:33]: LocId

File: Content.Shared/Damage/Systems/DamageOnInteractSystem.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+        if (totalDamage != null && totalDamage.AnyPositive())
+        {
+            _adminLogger.Add(LogType.Damaged, $"{ToPrettyString(args.User):user} injured their hand by interacting with {ToPrettyString(args.Target):target} and received {totalDamage.GetTotal():damage} damage");
+            args.Handled = true;
+
+            if (!_gameTiming.IsFirstTimePredicted || _net.IsServer)
+                return;
+
+            _audioSystem.PlayPvs(entity.Comp.InteractSound, args.Target);
```
Comment (Tayrtahn) [inline Content.Shared/Damage/Systems/DamageOnInteractSystem.cs:72]: PlayPredicted?

Comment (Plykiya) [reply Content.Shared/Damage/Systems/DamageOnInteractSystem.cs:72]: I did try that, it oddly didn't work, I can always try again?

Comment (Tayrtahn) [reply Content.Shared/Damage/Systems/DamageOnInteractSystem.cs:72]: Oh, I suppose the isServer check above takes care of it

**Comment by Plykiya** [comment]
checking if something is both powered and toggled on is kinda hard, in a generic system

my idea was using Delta's ComponentToggler to toggle whether or not the entity has the DamageOnInteractComponent, but it also means that whenever the lights flicker on or off, every single light on the station would have their component removed/added every single time

something tells me that's not really a good idea.......

**Comment by deltanedas** [comment]
the big prank would obsolete this :trollface:

```
- type: TriggerOnActivate
- type: Triggers
  groups:
    Activate:
      selectors:
      - !type:PoweredSelector
      - !type:UserSelector
      handlers:
      - type:DamageOnTrigger
        damage:
          types:
            Burn: 5
```

**Comment by Plykiya** [comment]
> the big prank would obsolete this :trollface:
> 
> ```
> - type: TriggerOnActivate
> - type: Triggers
>   groups:
>     Activate:
>       selectors:
>       - !type:PoweredSelector
>       - !type:UserSelector
>       handlers:
>       - type:DamageOnTrigger
>         damage:
>           types:
>             Burn: 5
> ```

how about the protection against that?

**Comment by deltanedas** [comment]
selector that filters user that doesnt have gloves on :trollface:

**Comment by Plykiya** [comment]
> selector that filters user that doesnt have gloves on :trollface:

hardsuits, eva suits, some random costumes that have gloves for some reason... guh

**Comment by Plykiya** [comment]
if the maintainers want this to be an easier to review PR I can split up the DamageOnInteraction and Lightbulb stuff into its own PR

File: Content.Shared/Damage/Systems/DamageOnInteractSystem.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                return;
+
+            _audioSystem.PlayPredicted(entity.Comp.InteractSound, args.Target, args.User);
+
+            if (_net.IsServer)
+                return;
+
+            if (entity.Comp.PopupText != null)
+                _popupSystem.PopupClient(Loc.GetString(entity.Comp.PopupText), args.User, args.User);
+        }
```
Comment (metalgearsloth) [inline Content.Shared/Damage/Systems/DamageOnInteractSystem.cs:81]: - PlayPredicted already checks IsFirstTimePredicted
- PopupClient doesn't actually do anything on server so checking for server first means it never does a popup.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**